### PR TITLE
Slots in rules are not validated against declared slots - [ENG 151]

### PR DESCRIPTION
**Pull request by varunshankar**
**Proposed changes**:
- Check unresolved slots before initiating model training.
- Collects all the unresolved slots and throws an error mentioning the same. Hence, allows user to know all the slot mismatches at the first try itself.
- Fixes: https://rasahq.atlassian.net/browse/ENG-151

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/model_training.py [hunk 1]
```diff
@@ -75,6 +79,22 @@ def _dry_run_result(
...
+        evnt.key
+        for step in stories.story_steps
+        for evnt in step.events
+        if isinstance(evnt, SlotSet)
+    ) - set(slot.name for slot in domain.slots)
+    if unresolved_slots:
+        rasa.shared.utils.cli.print_error_and_exit(
+            f"Unresolved slots found in stories/rules🚨 \n"
+            f"Tried to set non existent slots - {', '.join(unresolved_slots)}.\n"
+            f"Make sure you added all your slots to your domain file."
```
Comment (madebymagnolia) [inline rasa/model_training.py:90]: > "Make sure you added all your slots to your domain file."

I'd suggest amending this wording to:

> "Make sure you add all your slots to the domain file correctly."

As well as reminding users to add *all* of their slots to the domain file, adding "correctly" reminds users to check the spelling of them - a likely cause of this error.

**Review by twerkmeister — APPROVED** [review]
Looks good! Just a small suggestion. I could imagine the `check_unresolved_slots` method could be useful at other places too, where we might handle the results differently. You could make a function that only returns the unresolved slots for general usage. And then have another function here for the training start that checks unresolved slots and fails when there are some before the training.

**Comment by varunshankar** [comment]
> Looks good! Just a small suggestion. I could imagine the `check_unresolved_slots` method could be useful at other places too, where we might handle the results differently. You could make a function that only returns the unresolved slots for general usage. And then have another function here for the training start that checks unresolved slots and fails when there are some before the training.

Can you suggest me a place to keep the `check_unresolved_slots` function? I'm not able to find a suitable place for the function

File: rasa/model_training.py (hunk 1 above)
Comment (varunshankar) [reply rasa/model_training.py:90]: Thank you for your suggestions. Rephrased it accordingly

**Pull request merged**